codec/src/codec.rs

```diff
@@ -3,53 +3,122 @@
 
-    /// Encodes a value to bytes.
+    /// Encodes a value to a `BytesMut` buffer.
+    ///
+    /// Panics if the `write` implementation does not write the expected number of bytes.
+    ///
+    /// (Provided method).
     fn encode(&self) -> BytesMut {
         let len = self.len_encoded();
         let mut buffer = BytesMut::with_capacity(len);
         self.write(&mut buffer);
-        assert_eq!(buffer.len(), len);
+        assert_eq!(buffer.len(), len, "write() did not write expected bytes");
         buffer
     }
+}
 
-    /// Reads a value from a buffer, returning an error if there is an error while reading.
-    fn read(buf: &mut impl Buf) -> Result<Self, Error>;
+// Automatically implement `Encode` for types with a known size.
+// Otherwise, the type must define its own `len_encoded()` method.
+impl<T: EncodeFixed> Encode for T {
+    fn len_encoded(&self) -> usize {
+        Self::LEN_ENCODED
+    }
+}
 
-    /// Decodes a value from a buffer.
+/// Trait for types that can be decoded from a buffer, ensuring the entire buffer is consumed.
+pub trait Decode<Cfg = ()>: Read<Cfg> + Size {
+    /// Decodes a value from a buffer, ensuring the buffer is fully consumed.
+    ///
+    /// For types with a known size, this method first checks that the buffer has the expected size.
     ///
-    /// Returns an error if there is an error while decoding or if there is extra data remaining
-    /// after decoding the value from the buffer.
-    fn decode<B: Buf>(mut buf: B) -> Result<Self, Error> {
-        let result = Self::read(&mut buf);
+    /// (Provided method).
+    fn decode_cfg(mut buf: impl Buf, cfg: Cfg) -> Result<Self, Error> {
+        // If we can, before reading, check that the buffer has the expected size.
+        if let Some(size) = Self::FIXED_SIZE {
+            if buf.remaining() < size {
+                return Err(Error::EndOfBuffer);
+            }
+            if buf.remaining() > size {
+                return Err(Error::ExtraData(buf.remaining() - size));
+            }
+        }
+
+        let result = Self::read_cfg(&mut buf, cfg)?;
+
+        // Check that the buffer is fully consumed.
         let remaining = buf.remaining();
         if remaining > 0 {
             return Err(Error::ExtraData(remaining));
         }
-        result
+
+        Ok(result)
     }
 }
 
-/// Trait for types that have a fixed-length encoding
-pub trait SizedCodec: Codec {
-    /// The encoded length of this value.
+// Automatically implement `Decode` for types that implement `Read`.
+impl<Cfg, T: Read<Cfg>> Decode<Cfg> for T {}
+
+/// Trait for types that can be encoded and decoded.
+pub trait Codec<Cfg = ()>: Encode + Decode<Cfg> {}
+
+/// Automatically implement `Codec` for types that implement `Encode` and `Decode`.
+impl<Cfg, T: Encode + Decode<Cfg>> Codec<Cfg> for T {}
+
+/// Trait for types with a known, fixed encoded length.
+pub trait FixedSize: Size {
+    /// The length of the encoded value.
     const LEN_ENCODED: usize;
```

Comment:
I like the sound of `FIXED_SIZE` much better than `LEN_ENCODED` (more clearly implies this is the fixed const)...if we align on only one const here, I think we should think about `LEN_ENCODED` a tad more.

Reply:
`ENCODED_SIZE`?

Reply:
I want something that makes it clear this is fixed

Reply:
(and can't be confused with implemented `len_encoded` (or whatever we call that)

Reply:
Changed to `SIZE`